### ISSUE TITLE
SSO LTPA/JWT support authentication filter for Liberty 

### ISSUE THREAD

**Issue by rbacalzo** [OpenLiberty/open-liberty#10091]
WAD for SSO LTPA/JWT support authentication filter:
https://ibm.box.com/s/hv3ftnr3dsa294j37pu153b5jdyi8fk8

Doc: https://github.com/OpenLiberty/docs/issues/3801

**In This epic we are adding only the SSO LTPA/JWT support authentication filter.** 

Recently a few developers who had experienced issues with WebSphere Liberty SSO in the past were asked to provide input on a new Liberty LTPA authentication filter mechanism under consideration for development.

Below is a summary of the responses:

---
The LTPA authentication filter sounds like a good idea.  I think it should work similarly to the authentication filter used by OIDC (
https://www.ibm.com/support/knowledgecenter/SSEQTP_liberty/com.ibm.websphere.wlp.doc/ae/rwlp_auth_filter.html and https://www.ibm.com/support/knowledgecenter/SSEQTP_liberty/com.ibm.websphere.liberty.autogen.nd.doc/ae/rwlp_config_openidConnectClient.html).

But I would give this priority to the filter implementation: 
1) If there is a Bearer token on the HTTP Authorization header, then ignore LTPA cookie and do not attempt any LTPA authentication (this is really just a specific case of #2 below)
2) If there is a specific HTTP Header, then do not attempt any LTPA authentication
3) If there is a specific Cookie or a value for a cookie, then do not attempt any LTPA authentication
4) If there is a specific string in the HTTP Request URL (including request parameters), then do not attempt any LTPA authentication

In terms of APIs, these would be the most useful: 
1) Enance the OIDC PropagationHelper (https://www.ibm.com/support/knowledgecenter/en/SSEQTP_liberty/com.ibm.websphere.javadoc.liberty.doc/com.ibm.websphere.appserver.api.oidc_1.0-javadoc/com/ibm/websphere/security/openidconnect/PropagationHelper.html) to return a String version of the IdToken (e.g. getTokenString()).  If propagating an IdToken to a downstream server as a Bearer token on the http authorization header, this needs to be a string.  The current method getIdToken() returns a com.ibm.websphere.security.openidconnect.token.IdToken  object.  This is useless for token propagation since you can't get the String form of that token to propagate to a subsequent downstream servers.

2) A public method to validate an LTPA token. 

It would also be good to know a way of identifying tWAS and Liberty jar files required for builds when we do use specific Liberty or tWAS APIs in our code so that we can have the correct build dependencies.  Right now, we have to dig through Liberty and tWAS deployed directories to search for the jars containing the classes we need to compile our code, then manually copy those jars into our build tree.

---

I think we had less experience needing to manipulate the application of LTPA checks, and more with limitations around various SSO TAIs (i.e. SAML, SPNEGO) needing more advanced filtering behavior in cases of multiple IdP or some users with native login, some with SSO, so I hope I'm understanding the request.
 
Anyway, For LTPA filter ideas I think I would rank as follows
1.	One more more Request URL patterns
2.	Presence of specific cookies and values
3.	Presence of other tokens in the headers.. (for us, this is primarily SAMLv2)
4.	A range of ip addresses
5.	One or more remote ip address
6.	Some browser information (IE, Firefox, Safari) .. 
I would also think that it's very important to have a robust logical operations supported in whatever the filter mechanism (i.e. grouping of filter expressions with AND and OR logical operators, usage of NOT). I recall SAML sp.filter having some limitations around being all AND or all OR in certain expressions.
 
I would imagine it would be useful having a per-User API that allows the application to set a cookie so that the LTPA authentication could be by-passed next time if needed for that user ( a formal way for us to decide which users might have cookies that will be filtered out by #2 above )

**Comment by una-tapa** [comment]
This is from the stack product leads I had opportunities to work with.    This git issue include great ideas on filtering out ltpa token for easier security configuration .   Roger Bacalzo kindly  put together his ideas with  Brian Laskey's.  These are from real customers scenarios and their day-to-day experience developing on Liberty.

**Comment by garypicher** [comment]
Design review follow-up actions:

- Is similar support available for JWT?  If so, ensure that this support is consistent.  If not, come up with a plan to add similar support for JWT under this feature or a different feature.

**Comment by nickolas-potvin** [comment]
Would it be possible to also get these filters for custom TAIs?

**Comment by utle** [comment]
https://github.com/OpenLiberty/open-liberty/issues/12859

**Comment by utle** [comment]
@garypicher 
yes, we update the WAD for JWT SSO feature to support authentication filter

**Comment by utle** [comment]
Manual PB: 
https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_D1K7wO6rEeqT3K3gf94lOQ

**Comment by utle** [comment]
https://github.com/OpenLiberty/open-liberty/issues/13800
https://github.com/OpenLiberty/open-liberty/issues/13801
https://github.com/OpenLiberty/open-liberty/issues/13802
https://github.com/OpenLiberty/open-liberty/issues/13803
https://github.com/OpenLiberty/open-liberty/issues/13804

**Comment by NottyCode** [comment]
@utle do you have any reference to work items to track the documentation work for auth filters? In the design review we spoke about deficiencies in current documentation so I'd like to make sure we have documentation issues raised somewhere before I approve.

**Comment by utle** [comment]
Hi @NottyCode Here is the issue for auth filter description and sample. [https://github.com/OpenLiberty/open-liberty/issues/11741]
The Open Liberty Document also has an article for Authentication Filter.

**Comment by utle** [comment]
Feature Test Summary - https://github.com/OpenLiberty/open-liberty/issues/13804

**Comment by utle** [comment]
Hi @donbourne 

Serviceability Approval Comment - Please answer the following questions for serviceability approval:

UFO -- does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR? **Yes** Have these issues been addressed in the implementation? **Yes**

Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).
a) What problem paths were tested and demonstrated? User login succeeds with valid config, user login fails with bad config (and error messages created), Dynamic updates to correct/incorrect/different configuration (change from using ticketCache to keytab, update bad file name to good file name).

b) Who did you demo to? Nichole Stewart
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3? Yes

SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them. Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature? UFO marked for no SVT, but I had Nichole do my SVC demo.
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?

Which L2 / L3 queues will handle PMRs for this feature? Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it. Ask Don Bourne if you need links or more info. WAS L3: Identity Management

Does this feature add any new metrics or emit any new JSON events? If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs? No new metrics or JSON events

**Comment by chirp1** [comment]
The input that Ut provided for the ID issue looks good. Approving.

**Comment by utle** [comment]
GA Blog - https://github.com/OpenLiberty/open-liberty/issues/16552
